mediatek: mt7622: Fix ubi detection and add support fw_printsys for Xiaomi AX6S

remittor opened the pull request:
For a full description of the problem with the detection of partition ubi, look here: https://forum.openwrt.org/t/adding-openwrt-support-for-xiaomi-redmi-router-ax6s-xiaomi-router-ax3200/111085/1042

Problem:
If you specify flag "denx,fit" for partition "kernel", then the kernel tries to find rootfs in the same partition during boot. But in fact, the placement of rootfs is precisely determined by the name of another partition - "ubi".
And on some devices (there is another NAND chip), the "Denx search engine" manages to find rootfs at the end of partition "kernel". But in fact it is empty there.

Fix:
You just need to remove the "denx,fit" flag from partition "kernel". After that, the kernel will definitely search for rootfs in partition "ubi".

Another fixes:
1) Add support fw_printsys and fw_setsys.
2) Setting the nvram parameters necessary for recovery via UART.

hauke commented:
Please describe in the commit message why these changes are needed.
Please add line breaks after 72 characters into the commit message.

Ansuel commented:
@remittor can you rework the commit and drop the problem, fix and another fix think?

Would be good to have a linear commit description with imperative logic like: `If a "denx,fit" flag is specified ...` `Fix this by removing...` `While at it also fix...`

Neustradamus commented:
@remittor: Have you seen the @Ansuel comment?

remittor commented:
@Ansuel , Please provide the finished text, which I will just paste into the PR. Otherwise, it will drag on for a long time, because I see no reason to correct the current description.

xabolcs commented:
> Please provide the finished text, which I will just paste into the PR

I'm not **Ansuel**, but let me have a suggestion about commits.

### First commit
Contains only the change of `target/linux/mediatek/dts/mt7622-xiaomi-redmi-router-ax6s.dts`

```
mediatek: mt7622: fix rootfs/ubi detection for Xiaomi AX6S

Problem:
If you specify flag "denx,fit" for partition "kernel", then the kernel
tries to find rootfs in the same partition during boot. But in fact,
the placement of rootfs is precisely determined by the name of another
partition - "ubi".
And on some devices (there is another NAND chip), the "Denx search
engine" manages to find rootfs at the end of partition "kernel". But
in fact it is empty there.

Fix:
You just need to remove the "denx,fit" flag from partition "kernel".
After that, the kernel will definitely search for rootfs in partition
"ubi".
```

### Second commit
Contains the nvram related changes
* `target/linux/mediatek/mt7622/base-files/etc/init.d/bootcount`

```
mediatek: mt7622: set uboot params for UART recovery for Xiaomi AX6S

Like other on Xiaomi routers, the serial access is blocked in
different ways.

Bootloader won't accept any serial input unless "boot_wait" u-boot
environment variable is changed to "on".

Vendor firmware won't accept any serial input until "uart_en" is
set to "1".

Check for these options at boot and set if necessary to ensure
easier UART recovery.
```

### Third commit
Contains the `bdata` change.
* `package/boot/uboot-envtools/files/mediatek_mt7622`
```
uboot-envtools: add support fw_printsys and fw_setsys for Xiaomi AX6S

The system parameters are contained in the bdata partition.
This was done for the ramips Xiaomi routers in commit 6c7e337c80f9.
```

Ansuel commented:
Yes splitting this to 3 different commits is a good idea.

Anyway ideally patch description and title should be command or indication to what the system should do.

Something like "We have a fix and this is how we can fix it." became "Apply this change to fix this bug."
(from descriptive to imperative)

So ideally:

```
By specifying the flag "denx,fit" for partition "kernel", the kernel try to find rootfs in the same
partition during boot. Reality is that the placement of rootfs is precisely determined by the
name of another partition -"ubi".
It was also found that on some device (for example devices with NAND chips), the "Denx
search engine" manages to find roots at the end of partition "kernel", but such partition
doesn't exist and is empty there.

Fix this by removing the "denx,fit" flag from partition "kernel". With this change the original
behavior of searchif rootfs in partition "ubi" is restored. 
```

for second something like
```
Majority of Xiaomi routers have the serial access blocked in different ways.

Xiaomi bootloader won't accept any serial input unless "boot_wait" u-boot
envirorment variable is changed to "on"

Bootloader won't accept any serial input unless "boot_wait" u-boot
environment variable is changed to "on".

Vendor firmware won't accept any serial input until "uart_en" is
set to "1".

Check for these options at boot and set if necessary to ensure
easier UART recovery.
```

for third
```
System parameters are contained in the bdata partition, Add support
for fw_printsys and fw_setsys for Xiaomi AX6S following similar commits like
6c7e337c80f9.
```


Notice that they should be wrapped to 80 columns and feel free to ask for comments about these suggestions.

What do you think?

remittor committed:
mediatek: mt7622: fix rootfs/ubi detection for Xiaomi AX6S

By specifying the flag "denx,fit" for partition "kernel", the kernel
try to find rootfs in the same partition during boot. Reality is that
the placement of rootfs is precisely determined by the name of another
partition -"ubi".
It was also found that on some device (for example devices with NAND
chips), the "Denx search engine" manages to find roots at the end of
partition "kernel", but such partition doesn't exist and is empty
there.

Fix this by removing the "denx,fit" flag from partition "kernel". With
this change the original behavior of searchif rootfs in partition "ubi"
is restored.

Signed-off-by: Oleg S <remittor@gmail.com>

remittor committed:
mediatek: mt7622: set uboot params for UART recovery for Xiaomi AX6S

Majority of Xiaomi routers have the serial access blocked in different
ways.

Xiaomi bootloader won't accept any serial input unless "boot_wait"
u-boot envirorment variable is changed to "on"

Bootloader won't accept any serial input unless "boot_wait" u-boot
environment variable is changed to "on".

Vendor firmware won't accept any serial input until "uart_en" is
set to "1".

Check for these options at boot and set if necessary to ensure
easier UART recovery.

Signed-off-by: Oleg S <remittor@gmail.com>

remittor committed:
uboot-envtools: add support fw_printsys and fw_setsys for Xiaomi AX6S

System parameters are contained in the bdata partition, Add support
for fw_printsys and fw_setsys for Xiaomi AX6S following similar commits
like 6c7e337c80f9.

Signed-off-by: Oleg S <remittor@gmail.com>

dangowrt commented on target/linux/mediatek/mt7622/base-files/etc/init.d/bootcount:
```diff
@@ -7,5 +7,13 @@ boot() {
 		mtd erase senv || true
 		;;
+	xiaomi,redmi-router-ax6s)
+		local boot_wait=$( fw_printenv boot_wait | cut -d = -f 2 )
```
Use `fw_printenv -n boot_wait` instead of pipe to `cut` for all three calls retrieving variables from uboot environment.

981213 commented:
I dropped the change for ubi start address and picked the first commit from this PR.

Neustradamus commented:
@remittor: What is the situation of your PR?
I have seen your changes since last @Ansuel comment.
